### DISCUSSION TITLE
Layers

### DISCUSSION THREAD

**Discussion by GoodBoyDigital** [pixijs/pixijs#11064] [Ideas]
## **PixiJS Layer API Overview**

The PixiJS Layer API provides a way to control the **rendering order** of objects independently of their **logical parent-child relationships** in the scene graph. This allows developers to decouple how objects are transformed (via their logical parent) from how they are rendered on the screen.

---

### **Key Concepts**

1. **RenderLayers Control Rendering Order**:
   - RenderLayers define where in the render stack objects are drawn, but they do not affect an object’s transformations (e.g., position, scale, rotation) or logical hierarchy.
   - RenderLayers can be added anywhere in the scene graph.

2. **Logical Parenting Remains Unchanged**:
   - Objects still have a **logical parent** for transformations via `addChild`.
   - Assigning an object to a layer does not reparent it.

3. **Explicit Control**:
   - Developers assign objects to layers using `renderLayer.add` and remove them using `renderLayer.remove`.

---

#### **1. Creating a RenderLayer**
A `RenderLayer` is a lightweight object responsible for controlling render order. It has no children or transformations of its own but can be inserted anywhere in the scene graph to define its render position.

```ts
const layer = new RenderLayer();
app.stage.addChild(layer); // Insert the layer into the scene graph
```

---

#### **2. Adding Objects to a Layer**
Use `renderLayer.add` to assign an object to a layer. This overrides the object’s default render order defined by its logical parent.

```javascript
const rect = new PIXI.Graphics();
container.addChild(rect);    // Add to logical parent
layer.addToLayer(rect);      // Control render order via the layer
```

---

#### **3. Removing Objects from a Layer**
To stop an object from being rendered in the layer, use `removeFromLayer`.

```javascript
layer.remove(rect); // Stop rendering rect via the layer
```

When an object is removed from its logical parent (`removeChild`), it is automatically removed from the layer.

---

#### **4. Re-Adding Objects to Layers**
If an object is re-added to a logical parent, it does not automatically reassign itself to the layer. Developers must explicitly reassign it.

```javascript
container.addChild(rect);    // Logical parent
layer.add(rect);      // Explicitly reassign to the layer
```

---

#### **5. Layer Position in Scene Graph**
A layer’s position in the scene graph determines its render priority relative to other layers and objects. Layers can be inserted anywhere in the scene graph.

```javascript
const backgroundLayer = new RenderLayer();
const uiLayer = new RenderLayer();

app.stage.addChild(backgroundLayer);
app.stage.addChild(world);
app.stage.addChild(uiLayer);
```

---

### **Complete Example**

Here’s how the Layer API might be used in a real-world scenario:

```javascript
const container = new Container();

const shadowLayer = new RenderLayer();
const worldLayer = new RenderLayer();

container.addChild(shadowLayer);
container.addChild(worldLayer);

for(var i = 0; i < 100000; i++)
{
    const tree = new Container();
    const shadow = Sprite.from("shadow.png");
    const view = Sprite.from("tree.png");
    
    tree.addChild(shadow);
    tree.addChild(view);

    shadowLayer.add(shadow);
    worldLayer.add(view);

    container.addChild(tree);
}
```

---

### **Features**

1. **Independent Render Control**:
   - Objects rendered via a layer are independent of their logical parent’s render order.

2. **Simple Layer Management**:
   - Use `add` and `remove` for explicit control.
   - Layers have no children or transformations, making them lightweight and flexible.

3. **Automatic Cleanup**:
   - Removing an object from its parent (`removeChild`) automatically removes it from the layer.

4. **Dynamic Layers**:
   - Layers can be added / removed / repositions dynamically just like containers, allowing for flexible scene setups.

4. **Sorting of layer children**:
   - items in the layers render list can be sorted via `layerZindex` 
---

### **Feedback Questions**

1. **Ease of Understanding**:
   - Is the purpose of the Layer API clear (controlling render order vs. logical hierarchy)?
2. **API Usability**:
   - Does the `add` and `remove` pattern feel intuitive?
   - Would you prefer somthing like  `rect.addToRenderLayer(layer)` this too?
3. **Behavior Expectations**:
   - Is the automatic removal from the layer when calling `removeChild` expected and helpful?
4. **Advanced Use Cases**:
   - Do you foresee any challenges or limitations with this approach for more complex scenes?

**Comment by GoodBoyDigital** [comment]
You can test RenderLayers out here on [this branch](https://github.com/pixijs/pixijs/tree/feature/layers) 👍

**Comment by koteelok** [comment]
How about we use `associate` and `dissociate` for layers?

```ts
const container = new Container();

const shadowLayer = new RenderLayer();
const worldLayer = new RenderLayer();

container.addChild(shadowLayer);
container.addChild(worldLayer);

for(var i = 0; i < 100000; i++)
{
    const tree = new Container();
    const shadow = Sprite.from("shadow.png");
    const view = Sprite.from("tree.png");
    
    tree.addChild(shadow);
    tree.addChild(view);

    shadowLayer.associate(shadow);
    worldLayer.associate(view);

    container.addChild(tree);
}
```

**Reply by bigtimebuddy** [reply]
attach/detach is a little shorter, but same idea.

**Reply by danielbarion** [reply]
I still prefer the `add` & `remove` since other similar methods has a similar pattern for names.

**Reply by reececomo** [reply]
+1 for add/remove

**Comment by shirecoding** [comment]
Sounds good, though what would be the default render order if say some of the children are added to a render layer and some are not?

**Comment by shirecoding** [comment]
Also does it allow adding multiple containers to a single render layer? then the term 'associate' might not be that good as it does imply that the order of adding does not matter, addChild would make more sense if the order is considered

```
shadowLayer.associate(shadow);
shadowLayer.associate(shadow2);
shadowLayer.associate(shadow3);
shadowLayer.associate(shadow4);
```

Would this be possible?

I do see a symmetry with the current container system if this is possible

```
shadowContainer.addChild(shadow)
shadowContainer.addChild(shadow2)
shadowContainer.addChild(shadow3)

// Create a single shadowRenderLayer

shadowRenderLayer.addChild(shadow3) // i can switch the render order
shadowRenderLayer.addChild(shadow1)
shadowRenderLayer.addChild(shadow2)

parent.add(shadowContainer)
parent.add(shadowRenderLayer)
```

Though there is an ambiguity if i only add some of them to the container but not the renderlayer

```
shadowRenderLayer.addChild(shadow2)
// where is shadow1,3 in the render order?
```


One way to disambiguiate the 2 orders (transform and render) would be to combine both systems

```
shadowContainer.addChild(shadow, {renderLayer: 0)
shadowContainer.addChild(shadow2, {renderLayer: 1})
shadowContainer.addChild(shadow3) // if render layer is not provided, use the child hierarchy number ie 2
shadowContainer.addChild(shadow4, {renderLayer: 3})
```

**Reply by GoodBoyDigital** [reply]
Thanks for your feedback!

`Also does it allow adding multiple containers to a single render layer? then the term 'associate' might not be that good as it does imply that the order of adding does not matter, addChild would make more sense if the order is considered`

`Would this be possible?`
Yup! Though I guess we would recomend to use the zIndex to keep things simple! 

Order is considored, but keen to differencitate between addChild container and addChild layer as they both do very different things.

`Though there is an ambiguity if i only add some of them to the container but not the renderlayer` 

In this instance, shadow 1 and 3 would still be rendered in the container they belong too and shadow 2 would be rendered where the layer is.

`One way to disambiguiate the 2 orders (transform and render) would be to combine both systems`

I like the convenience here, let me think about how this could work but still feel 'pixi'.

**Reply by shirecoding** [reply]
hmm  @GoodBoyDigital  actually i do see a problem with my initial suggestion

```ts
// PROBLEM: renderLayer is global (cannot be local to parent)

parent.addChild(brother, {renderLayer: 1})  // eg. how to make renderLayer wrt greatGrandParent
grandParent.addChild(sister, {renderLayer: 2})
greatGrandParent.addChild(parent)
greatGrandParent.addChild(grandParent)


// Unless we can specify the parent then it might make sense
grandParent.addChild(sister, {renderLayer: 2, parent: greatGrandParent})
```

It sacrifices the locality of parent child relationship. eg, If the entire parent moves, the render layers of all its children moves with it.

For instance, I have an avatar in an isometric game (hands, legs, face, back hair, front hair) in 1 container, and within that container i want to set a custom zIndex for each component within the entire avatar container. But then when the avatar moves its own zIndex changes depending how far it is from the camera. In this case global zIndex would be a pain as i would need to not only update the avatar zIndex, but recalculate all the zIndex of the containers in the avatar.

The original RenderLayer() solves this. However i think it might be quite verbose though, would it be more convenient to add the method to the Container class instead?

```ts
anyContainer.addToRenderLayer(anyOtherContainer)
```

**Reply by GoodBoyDigital** [reply]
Is this issue related to z-sorting everything? For example, could character A's legs and face interact with character B's legs and face in such a way that parts of one character get mixed up in terms of depth?

**Reply by shirecoding** [reply]
Ah, just pointing out that there is a problem with my initial suggestion. I think the RenderLayer is still needed (your original approach) to be able to reorder locally within a container

ie. to keep the order within container. So that if i add another `container2` and add it to app.stage, the entire contents of `container` would be behind `container2` if it has a smaller zIndex

```ts
const container = new Container();

const shadowLayer = new RenderLayer();
const worldLayer = new RenderLayer();

container.addChild(shadowLayer);
container.addChild(worldLayer);
```

Though if Containers themselves were RenderLayers it would be quite more convenient to do

```
anyContainer.addToRenderLayer(anyOtherContainer)
```

instead of creating and managing renderlayers and adding them to containers. Something like this

```ts
const avatar = new Container()
const head = new Container()
const backHair = new Container()
const frontHair = new Container()
const torso = new Container()

head.addChild(backHair)
head.addChild(frontHair)
torso.addChild(head)
avatar.addChild(torso)

avatar.addToRenderLayer(backHair)
avatar.addToRenderLayer(head)
avatar.addToRenderLayer(torso)
avatar.addToRenderLayer(frontHair)
```

**Reply by GoodBoyDigital** [reply]
my brain is not working today so you will have to bare with me :P

could you show me what that last bit of code would look like if we have access to `anyContainer.addToRenderLayer(anyOtherContainer)` 

thanks!

**Reply by shirecoding** [reply]
i guess it would just be like this

```ts
const avatar = new Container()
const head = new Container()
const backHair = new Container()
const frontHair = new Container()
const torso = new Container()

head.addChild(backHair)
head.addChild(frontHair)
torso.addChild(head)
avatar.addChild(torso)

// In this implementation we don't need to create a seperate RenderLayer and add it to avatar
avatar.addToRenderLayer(backHair)
avatar.addToRenderLayer(head)
avatar.addToRenderLayer(torso)
avatar.addToRenderLayer(frontHair)
```

This would require adding the RenderLayer in the Container class though. or could be lazily created on the call to Container.addToRenderLayer

vs

```ts

const avatar = new Container()
const head = new Container()
const backHair = new Container()
const frontHair = new Container()
const torso = new Container()

head.addChild(backHair)
head.addChild(frontHair)
torso.addChild(head)
avatar.addChild(torso)

// A little more verbose here as we create renderLayer as an intermediary and add it into the container
const renderLayer = new RenderLayer()
avatar.addChild(renderLayer)

renderLayer.addToRenderLayer(backHair)
renderLayer.addToRenderLayer(head)
renderLayer.addToRenderLayer(torso)
renderLayer.addToRenderLayer(frontHair)
```

**Reply by shirecoding** [reply]
Hmm on second thoughts, i think it only makes it slightly more convenient in my use case (to use containers as render layers).

but it doesnt really do much in this example, but maybe muddies the naming a little more

```ts
const container = new Container();
const shadowLayer = new RenderLayer();
const worldLayer = new RenderLayer();

container.addChild(shadowLayer);
container.addChild(worldLayer);

for(var i = 0; i < 100000; i++)
{
    const tree = new Container();
    const shadow = Sprite.from("shadow.png");
    const view = Sprite.from("tree.png");
    
    tree.addChild(shadow);
    tree.addChild(view);

    shadowLayer.add(shadow);
    worldLayer.add(view);

    container.addChild(tree);
}
```

vs

```ts
const container = new Container();
const shadowContainer = new Container();
const worldContainer = new Container();

container.addChild(shadowContainer);
container.addChild(worldContainer);

for(var i = 0; i < 100000; i++)
{
    const tree = new Container();
    const shadow = Sprite.from("shadow.png");
    const view = Sprite.from("tree.png");
    
    tree.addChild(shadow);
    tree.addChild(view);

    shadowContainer.addToRenderLayer(shadow);
    worldContainer.addToRenderLayer(view);

    container.addChild(tree);
}
```

Hmm i guess its better to stick to the `new RenderLayer()` for flexibility!

**Reply by GoodBoyDigital** [reply]
cool! I agree having `RenderLayer` allows the dev to place it in the scene grapch to manage the ordering themselfs a little easier

**Comment by leighcarey-playco** [comment]
> Does the add and remove pattern feel intuitive?

I think `add` and `remove` are clear - no need to overcomplicate the naming. Although one question is what happens when you add an object to a layer when it's already part of a different layer? I expect an object can be added to at most one layer, but is it automatically removed from its current layer, or is this an error scenario?

Along these lines, is it possible to get a reference to an object's current layer? The use-case here is for temporarily promoting an object to a "higher" layer, and then setting it back again. How would we achieve this?

> Would you prefer somthing like rect.addToRenderLayer(layer) this too?

Restricting the render layer API to the `RenderLayer` class is preferable here imo. If we need a reference to both the object and the render layer to change the render order, there's no advantage to having an `.addToRenderLayer` mixin or similar.

**Reply by GoodBoyDigital** [reply]
thanks for the feedback @leighcarey-playco !

- You assumption is correct. I item can only exist in one layer - it will be removed from the layer it is currently attached to before being attached to a new layer.

- You would be able to access the parent render layer via the prop `parentRenderLayer`

- Agree, things could get messy if the surface area of the API gets too big!

**Comment by StepanMynarik** [comment]
@GoodBoyDigital Is it still going to be possible to render individual layers to a render texture?

I'm currently trying to find a way on how to implement a deferred rendering in Pixi v8.

**Reply by GoodBoyDigital** [reply]
thats a great question - Perhaps not directly - but at the very least you will be able to add a renderLayer to a container and render that container.

**Comment by Flippsor** [comment]
Sounds like great news!

Not quite sure if the automatic cleanup on removeChild is a good idea. What if a child just gets a new parent? Then we have to reassign the Layer again. Was this implemented in the Pixi-Layer Plugin, too?

Is there already a planned release date for this feature/branch?

**Comment by reececomo** [comment]
Just for the uninitiated --> what's a real-world situation that you need layers > containers? Do other rendering engines / sprite managers have some kind of similar concept?

Only asking because 2D renderings been around for a long time and Containers + zIndex seems pretty intuitive to me for organizing layers, so adding a new concept to the core library _also_ called layers adds a bit of ambiguity for me.

**Reply by danielbarion** [reply]
I have a world container and a character container, my world container has a lot of containers (one for each X,Y because it can contain multiple sprites in order of different Z levels), so to keep it organized, I'm not merging them into a flat container, also if I had to do that, I will need to refactor a lot my code, but this blocks me today to move the character behind the top of the three, or behind some object, the layers will solve this issue.

**Reply by reececomo** [reply]
I didn’t quite follow that example… how does layers solve that?

**Reply by reececomo** [reply]
Having zIndex be global would be interesting (resolving globally, then locally) — if I recall this is how a few other engines do it.

Having two independent systems for managing z depth might sound like a simple fix here and might be suitable for a plugin, but bringing it into the core engine as-is may hurt more than it helps 🤷‍♀️

**Reply by shirecoding** [reply]
You can look at Spine2D for a good example, they implemented Slots for the very reason to separate the transform parent vs the render parent.

For example, If you have a "long hair" which you make the transform parent "head", but "front hand" is in front of "head",  then you want your "long hair" to appear in front of your "front hand". It is impossible in this case because you cant change the render order of the "long hair" to be in front of "front hand" as it's parent is the "head" which is below "front hand"

**Comment by Flippsor** [comment]
Tried out the new Render Layers today and here some feedback:

1. Automatically removing the child from the RenderLayer is NOT a good solution in my case. I have a CAD like application where a lot of reparenting of different containers is done (i.e. grouping, moving containers under other containers). In this case we have to check not only the container which was moved but ALL child container of this container and reattach the layer again. Retriggering the attachement is quite complex to do. 
So in my case it would be way easer to deactivate the automatical detachment (by a setting) and do it myself.

2. Changing the transformation (x,y, rotation) of Containers in a Layer seem to move with respect to the RenderLayer coordinate system and NOT in the parent coordinate system. This seems like a bug but i have test this further. Right now i can just say that the transformation does behave different with RenderLayer (which should not change)
